src/main/java/javaslang/collection/List.java

```diff
@@ -695,7 +701,18 @@ public Short next() {
     @Override
     default List<T> filter(Predicate<? super T> predicate) {
         Objects.requireNonNull(predicate, "predicate is null");
-        return isEmpty() ? this : foldLeft(List.<T> empty(), (xs, x) -> predicate.test(x) ? xs.prepend(x) : xs).reverse();
+        List<T> list = this;
+        List<T> result = List.empty();
+        boolean testFailed = false;
+        while (!list.isEmpty()) {
+            if (predicate.test(list.head())) {
+                result = result.prepend(list.head());
+            } else {
+                testFailed = true;
+            }
+            list = list.tail();
+        }
+        return testFailed ? result.reverse() : this;
```

Comment:
Mmhh, we need to iterate all elements. Me may save the last reverse() even shorter without variables:

```java
final List<T> filtered = foldLeft(List.<T> empty(), (xs, x) -> predicate.test(x) ? xs.prepend(x) : xs);
return (this.length() == filtered.length()) ? this : filtered.reverse();
```